File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
# DFS와 BFS
# 그래프 기초
# dfs 알고리즘은 스택과 재귀함수를 사용하여 구현하고, 
# bfs 알고리즘은 while문과 deque를 이용하여 구현하자.(시간복잡도 낮은 popleft를 사용하자)

from collections import deque
n, m, v = map(int, input().split())

graph = [[False] * (n+1) for _ in range(n+1)]

for i in range(m):
    a, b = map(int, input().split())
    graph[a][b] = 1
    graph[b][a] = 1

visited1 = [False] * (n+1)
visited2 = [False] * (n+1)

def dfs(v):
    visited1[v] = True
    print(v, end=' ')
    
    for i in range(1, n+1):
        if visited1[i] == False and graph[v][i] == 1:
            visited1[i] = True
            dfs(i)

def bfs(v):
    q = deque([v]) # 첫번째꺼 넣고 시작
    visited2[v] = True
    while(q): # 데크에 값 없을 때까지 계속 돌아
        v = q.popleft()
        print(v, end=' ')
        for i in range(1, n+1):
            if visited2[i] == False and graph[v][i] == 1:
                q.append(i)
                visited2[i] = True          
dfs(v)
print()
bfs(v)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
// println 대신 string builder 쓰도록 바꿈
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Main {
    static StringBuilder sb = new StringBuilder();
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int V = Integer.parseInt(st.nextToken());

        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph.get(a).add(b);
            graph.get(b).add(a);
        }

        for (ArrayList<Integer> a : graph) {
            Collections.sort(a);
        }

        boolean[] visited1 = new boolean[N + 1];
        boolean[] visited2 = new boolean[N + 1];

        dfs(graph, V, visited1);
        sb.append("\n");
        bfs(graph, V, visited2);
        System.out.println(sb);

    }

    private static void dfs(ArrayList<ArrayList<Integer>> graph, int v, boolean[] visited1) {
        visited1[v] = true;
        sb.append(v).append(" ");
        for (int n : graph.get(v)) {
            if (!visited1[n]) {
                dfs(graph,n,visited1);
            }
        }
    }

    private static void bfs(ArrayList<ArrayList<Integer>> graph, int v, boolean[] visited2) {
        Queue<Integer> q = new LinkedList<>();
        visited2[v] = true;
        q.offer(v);
        while (!q.isEmpty()) {
            int n = q.poll();
            sb.append(n).append(" ");
            for (int node : graph.get(n)) {
                if (!visited2[node]) {
                    visited2[node] = true;
                    q.offer(node);
                }
            }
        }
    }


}